sip/sip-002-smart-contract-language.md

```diff
@@ -98,48 +98,85 @@ any type.
-smart contract identifier is a _principal_.
+`(contract-call?)` function.
 
-For example, to call the function `register-name` in a smart contract,
-you would use:
+This function returns a Response type result-- the return value of the called smart
+contract function. Note that if a called smart contract returns an
+`err` type, it is guaranteed to not alter any smart contract state
+whatsoever. Of course, any transaction fees paid for the execution
+of that function will not be returned.
+
+We distinguish 2 different types of `contract-call?`:
+
+* Static dispatch: the callee is a known, invariant contract available
+on-chain when the caller contract is being deployed. In this case, the
+callee's principal is provided as first argument, followed by the name
+of the method and its arguments:
 
 ```scheme
-(contract-call!
-    'SC3H92H297DX3YDPFHZGH90G8Z4NPH4VE8E83YWAQ
-    'register-name
+(contract-call?
+    'SC3H92H297DX3YDPFHZGH90G8Z4NPH4VE8E83YWAQ.registrar
+    register-name
     name-to-register)
 ```
 
-This function returns a boolean-- the return value of the called smart
-contract function. Note that if a called smart contract returns an
-`err` type, it is guaranteed to not alter any smart contract state
-whatsoever. Of course, any transaction fees paid for the execution
-of that function will not be returned.
+This approach must always be preferred, when adequat.
+It makes static analysis easier, and eliminates the
+potential for reentrancy bugs when the contracts are
+being published (versus when being used).
+
+* Dynamic dispatch: the callee is passed as an argument, and typed
+as a trait reference (<A>).
+
+```scheme
+(define-public (swap (token-a <can-transfer-tokens>)
+                     (amount-a uint)
+                     (owner-a principal)
+                     (token-b <can-transfer-tokens>)
+                     (amount-b uint)
+                     (owner-b principal)))
+     (begin
+         (unwrap! (contract-call? token-a transfer-from? owner-a owner-b amount-a))
+         (unwrap! (contract-call? token-b transfer-from? owner-b owner-a amount-b))))
+```
+
+Traits can either be locally defined:
+
+```scheme
+(define-trait can-transfer-tokens (
+    (transfer-from? (principal principal uint) (response uint)))
+```
+
+Or imported from an existing contract:
+
+```scheme
+(use-trait can-transfer-tokens
+    'SC3H92H297DX3YDPFHZGH90G8Z4NPH4VE8E83YWAQ.contract-defining-trait.can-transfer-tokens)
+```
+
+Looking at trait conformance, callee contracts have two different paths.
+They can either be "compatible" with a trait by defining methods
+matching some of the methods defined in a trait, or explicitely declare
+conformance using the `impl-trait` statement:
+
+```scheme
+(impl-trait 'SC3H92H297DX3YDPFHZGH90G8Z4NPH4VE8E83YWAQ.contract-defining-trait.can-transfer-tokens)
+```
+
+Explicit conformance should be prefered when adequat.
+It acts as a garde-fou and help the static analysis detecting
```

Comment:
I think "garde-fou" translates to "safeguard" here.

"It acts as a safeguard by helping the static analysis system to detect deviations in method signatures before contract deployment."

Reply:
Addressed with https://github.com/blockstack/blockstack-core/pull/1246/commits/be3a2497635959001639308e52364ad6876a168c